server/jetstream_api.go

```diff
@@ -2341,6 +2361,92 @@ func (s *Server) jsLeaderServerStreamMoveRequest(sub *subscription, c *client, _
 
 	cfg.Placement = origPlacement
 
+	s.Noticef("Requested move of: R=%d stream %s in account %s from old peer set %+v to new peer set %+v",
+		cfg.Replicas, streamName, accName, s.peerSetToNames(currPeers), s.peerSetToNames(peers))
+
+	// execute on a go routine for route or gateway, as this may block waiting from responses from other servers and
+	// therefore block all route/gateway traffic.
+	if c.kind != ROUTER && c.kind != GATEWAY {
+		go s.jsClusteredStreamUpdateRequest(&ciNew, targetAcc.(*Account), subject, reply, rmsg, &cfg, peers)
```

Comment:
Need to make sure you have copies since you are running that in go routine. So for sure, replace `rmsg` with `copyBytes(rmsg)`. Could you do the same in existing `jsStreamUpdateRequest()`?

Reply:
That copy does not need to be made


This is the code where I initially copied this from.
It does not copy either.
```
func (s *Server) jsStreamUpdateRequest(sub *subscription, c *client, _ *Account, subject, reply string, rmsg []byte) {
	...
	// Handle clustered version here.
	if s.JetStreamIsClustered() {
		// If we are inline with client, we still may need to do a callout for stream info
		// during this call, so place in Go routine to not block client.
		if c.kind != ROUTER && c.kind != GATEWAY {
			go s.jsClusteredStreamUpdateRequest(ci, acc, subject, reply, rmsg, &cfg, nil)
		} else {
			s.jsClusteredStreamUpdateRequest(ci, acc, subject, reply, rmsg, &cfg, nil)
		}
		return
	}
```

I believe the copy in apiDispatch is the reason the rmsg is not being copied in jsStreamUpdateRequest.

```
func (js *jetStream) apiDispatch(sub *subscription, c *client, acc *Account, subject, reply string, rmsg []byte) {
	// No lock needed, those are immutable.
	s, rr := js.srv, js.apiSubs.Match(subject)
...

	// If this is directly from a client connection ok to do in place.
	if c.kind != ROUTER && c.kind != GATEWAY {
		jsub.icb(sub, c, acc, subject, reply, rmsg)
		return
	}

	// If we are here we have received this request over a non client connection.
	// We need to make sure not to block. We will send the request to a long-lived
	// go routine.

	// Copy the state. Note the JSAPI only uses the hdr index to piece apart the
	// header from the msg body. No other references are needed.
	s.jsAPIRoutedReqs.push(&jsAPIRoutedReq{jsub, sub, acc, subject, reply, copyBytes(rmsg), c.pa})
}
```
As jsStreamUpdateRequest is called via apiDispatch

```
	// API handles themselves.
	pairs := []struct {
		subject string
		handler msgHandler
	}{
...
		{JSApiStreamUpdate, s.jsStreamUpdateRequest},
...
	}

	js.mu.Lock()
	defer js.mu.Unlock()

	for _, p := range pairs {
		sub := &subscription{subject: []byte(p.subject), icb: p.handler}
		if err := js.apiSubs.Insert(sub); err != nil {
			return err
		}
	}
```

Reply:
Well, if this is invoked by apiDispatch and we are already making a copy and routing to a single go routine, then we should not "go" run that again here, don't you think?

But jsLeaderServerStreamMoveRequest itself seem to be attached directly to a sys sub, and then you MUST make a copy:
```
	if cc.peerStreamMove == nil {
		cc.peerStreamMove, _ = s.systemSubscribe(JSApiServerStreamMove, _EMPTY_, false, c, s.jsLeaderServerStreamMoveRequest)
	}
```